### PR TITLE
mesa: add 19.1 recipe

### PR DESCRIPTION
**Pull request by kenmays**
Provides Mesa 19.,1.7 recipe and patchset for Haiku - tested on hrev53452 x86_64 w/Python 3.7.3.

NOTE: Recipe disabled for all builds by default. Not for use with gcc2. Safe to merge. (Packages built for Haiku). Current Haiku pthread implementation/workarounds may limit future Mesa 19.2/19,3+ integrations. 

Use only for dev/testing purposes.

### CODE REVIEW

**Comment by extrowerk** [comment]
```
meson.build:775:2: ERROR: Problem encountered: Python (3.x) mako module >= 0.8.0 required to build mesa.
```

**Comment by extrowerk** [comment]
```
../src/gallium/state_trackers/hgl/hgl.c:48:1: error: static declaration of 'hgl_st_framebuffer' follows non-static declaration
 hgl_st_framebuffer(struct st_framebuffer_iface *stfbi)
```

**Comment by kenmays** [comment]
[mesa-19.1.7_build-log-Haiku_haikuporter-install-review_cocobean.txt](https://github.com/haikuports/haikuports/files/3634304/mesa-19.1.7_build-log-Haiku_haikuporter-install-review_cocobean.txt)

Build log. Everything builds. Polishing up installation.

File: sys-libs/mesa/patches/mesa-19.1.7.patchset [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+ GL_API void GL_APIENTRY glPrimitiveBoundingBox (GLfloat minX, GLfloat minY, GLfloat minZ, GLfloat minW, GLfloat maxX, GLfloat maxY, GLfloat maxZ, GLfloat maxW);
+ 
++GL_API void GL_APIENTRY glBlendBarrier (void);
++GL_API void GL_APIENTRY glPrimitiveBoundingBox (GLfloat minX, GLfloat minY, GLfloat minZ, GLfloat minW, GLfloat maxX, GLfloat maxY, GLfloat maxZ, GLfloat maxW);
++
+ /* Enable frame pointer elimination on Windows, otherwise forgetting to add
+  * APIENTRY to _mesa_* entrypoints will not cause crashes on debug builds, as
+  * the initial ESP value is saved in the EBP in the function prologue, then
+
+diff --git a/src/gallium/state_trackers/hgl/Makefile.am b/src/gallium/state_trackers/hgl/Makefile.am
```
Comment (extrowerk) [inline sys-libs/mesa/patches/mesa-19.1.7.patchset]: Umm, MESA switched from auto* to Meson, all the Makefiles should be rewritten for meson.

File: sys-libs/mesa/patches/mesa-19.1.7.patchset [hunk 2]
```diff
@@ -0,0 +1,414 @@
+From f632aac9385cfe02f5bfe762d93e60b5b46f2c77 Tue Sep 17 00:00:00 2019
+From: Ken Mays <kmays2000@gmail.com>
```
Comment (extrowerk) [inline sys-libs/mesa/patches/mesa-19.1.7.patchset]: AFAIk your patchset based on he 17.x patchset, if so,be respectful and do not takeover their work.

**Comment by extrowerk** [comment]
> [](https://github.com/haikuports/haikuports/files/3634304/mesa-19.1.7_build-log-Haiku_haikuporter-install-review_cocobean.txt)
No mako package attached to the chroot. How is meson not complaining about it?

>Installing src/mapi/shared-glapi/libglapi.so.0.0.0 to /usr/local/lib
>Installing src/mapi/es1api/libGLESv1_CM.so.1.1.0 to /usr/local/lib
>Installing src/mapi/es2api/libGLESv2.so.2.0.0 to /usr/local/lib

The install path is not correct, otherwise the recipe doesn't defines this libs as provides. Are they required for Haiku? AFAIK the current (17.x) package doesn't contains them. Where did they come from?

>../src/gallium/auxiliary/rbug/rbug_context.c:214:10: warning: integer overflow in expression '-2147483648' of type 'int' results in '-2147483648' [-Woverflow]
This seems to be suspicious. Haiku uses negative errors.

The attached log is incomplete, we would need a complete buildlog to see, what's gets built and what not. Care to provide a complete buildlog? If so, please backup your workfolder and let haikuporter build it from scratch.

**Comment by kenmays** [comment]
Fixed install. Packages built:
waiting for build package mesa-19.1.7-1 to be deactivated
waiting for build package mesa_debuginfo-19.1.7-1 to be deactivated
waiting for build package mesa_devel-19.1.7-1 to be deactivated
waiting for build package mesa_swpipe-19.1.7-1 to be deactivated
grabbing mesa-19.1.7-1-x86_64.hpkg and moving it to /boot/home/haikuports/packages/mesa-19.1.7-1-x86_64.hpkg
grabbing mesa_debuginfo-19.1.7-1-x86_64.hpkg and moving it to /boot/home/haikuports/packages/mesa_debuginfo-19.1.7-1-x86_64.hpkg
grabbing mesa_devel-19.1.7-1-x86_64.hpkg and moving it to /boot/home/haikuports/packages/mesa_devel-19.1.7-1-x86_64.hpkg
grabbing mesa_swpipe-19.1.7-1-x86_64.hpkg and moving it to /boot/home/haikuports/packages/mesa_swpipe-19.1.7-1-x86_64.hpkg
~/haikuports/sys-libs/mesa>

File: sys-libs/mesa/patches/mesa-19.1.7.patchset (hunk 2 above)
Comment (pulkomandy) [reply sys-libs/mesa/patches/mesa-19.1.7.patchset]: We don't use that authorship information anywhere, and this is the default behavior of --extract-patchset. I don't see why we should nitpick on this. There are enough real problems to fix here.

**Comment by extrowerk** [comment]
For me it still fails:
```
../src/gallium/targets/haiku-softpipe/GalliumContext.cpp: In member function 'status_t GalliumContext::SwapBuffers(context_id)':
../src/gallium/targets/haiku-softpipe/GalliumContext.cpp:343:44: error: 'struct st_context::<unnamed>' has no member named 'framebuffer'
  unsigned nColorBuffers = stContext->state.framebuffer.nr_cbufs;
                                            ^~~~~~~~~~~
../src/gallium/targets/haiku-softpipe/GalliumContext.cpp:345:44: error: 'struct st_context::<unnamed>' has no member named 'framebuffer'
   pipe_surface* surface = stContext->state.framebuffer.cbufs[i];
                                            ^~~~~~~~~~~
```

**Comment by extrowerk** [comment]
After enabling the patchset, it built but packaging fails in strict mode:
```
Warning: POLICY ERROR: Invalid top-level package entry "share"

```

**Comment by extrowerk** [comment]
After installing the generated package:
```
~> GLInfo 
runtime_loader: Static TLS model is not supported.
runtime_loader: /boot/system/lib/libGL.so: Troubles handling dynamic section
```

File: sys-libs/mesa/patches/mesa-19.1.7.patchset (hunk 2 above)
Comment (korli) [reply sys-libs/mesa/patches/mesa-19.1.7.patchset]: > We don't use that authorship information anywhere, and this is the default behavior of --extract-patchset. I don't see why we should nitpick on this. There are enough real problems to fix here.

IMO regarding upstream, this makes things more complex for no good reason. This patch is probably useless anyway.

File: sys-libs/mesa/patches/mesa-19.1.7.patchset (hunk 1 above)
Comment (kenmays) [reply sys-libs/mesa/patches/mesa-19.1.7.patchset]: Fixed.

File: sys-libs/mesa/patches/mesa-19.1.7.patchset (hunk 2 above)
Comment (kenmays) [reply sys-libs/mesa/patches/mesa-19.1.7.patchset]: @korli - The BGLView patch was needed from the 'Flurry' and 'Gravity' screensaver bug fix.
The other patches fix the native build issue. Mainly, 2-4 patches remain now which I upstreamed for Mesa-dev review.

**Comment by kenmays** [comment]
I reported everything upstream (only three Haiku-specific patches remain) and provided my build log to the Mesa-dev team. Got feedback. Hope to see this in the 19.1.8 and 19.2 stable releases.

Thanks everyone!!

File: sys-libs/mesa/patches/mesa-19.1.7.patchset (hunk 2 above)
Comment (korli) [reply sys-libs/mesa/patches/mesa-19.1.7.patchset]: Looks fun indeed: https://gitlab.freedesktop.org/mesa/mesa/merge_requests/2062

**Comment by kenmays** [comment]
Submitted HGL-related patches were accepted by Mesa-dev for mesa-stable of 09/26/2019. Closing this ticket.

**Comment by korli** [comment]
well the MR is still to be merged and points to master (19.3-devel), not stable.

**Pull request closed without merging**